### PR TITLE
Port codebase to TypeScript 🧪

### PR DESCRIPTION
**Pull request by jpmonette**
As we are heavily relying on the MobileSDK for an internal project, we've initially worked on adding all types using a custom type definition, and now on a port of the codebase into TypeScript to make type checking and further addition of features easier.

This is still a WIP we are sharing, if you have any input and see your general appetite for the port.

Resolves #185

---

Outstanding tasks:
- [x] Complete port of the business logic
- [x] Add missing types
- [x] Migrate the tests + test + make tests pass

### CODE REVIEW

File: src/react.force.smartstore.ts [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+  exec(successCB, errorCB, "upsertSoupEntries", {
+    soupName,
+    entries,
+    externalIdPath,
+    isGlobalStore: storeConfig.isGlobalStore,
+    storeName: storeConfig.storeName,
+  });
+};
+
+export const removeFromSoup = <T>(
```
Comment (trooper2013) [inline src/react.force.smartstore.ts:601]: Returns a string "OK" on success.

File: src/react.force.smartstore.ts [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+
+  exec(successCB, errorCB, "closeCursor", {
+    cursorId: cursor.cursorId,
+    isGlobalStore: storeConfig.isGlobalStore,
+    storeName: storeConfig.storeName,
+  });
+};
+
+// ====== Store Operations ======
+export const getAllStores = <T>(
```
Comment (trooper2013) [inline src/react.force.smartstore.ts:718]: Returns an array of dictionaries. [ {storeName:'' , isGlobalStore: true or false},{}]

File: src/react.force.smartstore.ts [hunk 3]
```diff
@@ -0,0 +1,757 @@
...
+  successCB: (result: T) => void,
+  errorCB: (err: Error) => void
+): void => {
+  exec(successCB, errorCB, "getAllStores", {});
+};
+
+export const getAllGlobalStores = <T>(
```
Comment (trooper2013) [inline src/react.force.smartstore.ts:725]: Returns an array of dictionaries. [ {storeName:'' , isGlobalStore: true or false},{}]

File: src/react.force.smartstore.ts [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+): void => {
+  storeConfig = checkFirstArg(storeConfig);
+
+  exec(successCB, errorCB, "removeStore", {
+    isGlobalStore: storeConfig.isGlobalStore,
+    storeName: storeConfig.storeName,
+  });
+};
+
+export const removeAllGlobalStores = <T>(
```
Comment (trooper2013) [inline src/react.force.smartstore.ts:745]: Returns a string "OK" on success.

File: src/react.force.smartstore.ts [hunk 5]
```diff
@@ -0,0 +1,757 @@
...
+  successCB: (result: T) => void,
+  errorCB: (err: Error) => void
+): void => {
+  exec(successCB, errorCB, "removeAllGlobalStores", {});
+};
+
+export const removeAllStores = <T>(
```
Comment (trooper2013) [inline src/react.force.smartstore.ts:752]: Returns a string "OK" on success.

File: src/typings/mobilesync.d.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+export type SyncDownTarget = {
+  type: "soql" | "sosl" | "mru" | "custom";
+  query: string;
+  modificationDateFieldName?: string;
+  iOSImpl?: string;
+  idFieldName?: string;
+};
+
+export type SyncUpTarget = {
+  createFieldlist?: Array<string>;
```
Comment (trooper2013) [inline src/typings/mobilesync.d.ts:28]: ``` createFieldlist?: string[] ```

File: src/typings/mobilesync.d.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  updateFieldlist?: Array<string>;
+};
+
+export type SyncOptions = {
+  mergeMode?: string;
+  fieldlist?: Array<string>;
```
Comment (trooper2013) [inline src/typings/mobilesync.d.ts:34]: same as above

File: src/react.force.mobilesync.ts [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+  target: SyncDownTarget,
+  soupName: string,
+  options: SyncDownOptions,
+  ...args: unknown[]
+): void => {
+  storeConfig = checkFirstArg(storeConfig);
+
+  let errorCB: (err: Error) => void;
+  let successCB: (result: SyncEvent) => void;
+  let syncName: string | undefined;
```
[the remark this answers is not in the document]
Comment (jpmonette) [reply src/react.force.mobilesync.ts:89]: @trooper2013 Isn't it preferable to leave it unassigned /`undefined` than having to explicitly set `null`?

**Comment by jpmonette** [comment]
Thanks for this initial review @trooper2013.

About testing it - this is still a WIP, but will give it a shot once I have a chance. As our project depending on this is closed source, I won't be able to share what we are currently using but will see if I can put something together.

Few things I have on top of my mind as I reviewed the codebase:
- Handling typings with `promiser` seems like a challenge. Could this be a good opportunity to officially return promises, similarly to how `jsforce` does it?
- Unless I misunderstood how this callback works (https://github.com/jpmonette/SalesforceMobileSDK-ReactNative/blob/dev/src/react.force.smartstore.ts#L662), it seems like `errorCB()` can either take an `Error` as first argument or a `StoreCursor`. I'd suggest to change it for consistency.
- Seems like `sendRequest` cannot be effectively promisified as the callbacks are passed in between a list of other arguments (https://github.com/jpmonette/SalesforceMobileSDK-ReactNative/blob/dev/src/react.force.net.ts#L54)
   - I feel a lot of functions could benefit from having a single object as an argument instead of many separate arguments - it would also simplify type checking.

File: src/react.force.common.ts [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+ * @template T
+ * @param {string} str
+ * @returns {T}
+ */
+export const safeJSONparse = <T>(str: string): T => {
+  try {
+    return JSON.parse(str);
+  } catch (e) {
+    // @ts-ignore
+    return str;
```
[the remark this answers is not in the document]
Comment (jpmonette) [reply src/react.force.common.ts:147]: @trooper2013 Is it possible to make sure that all that the Android bridge would return is an object for consistency? I do not know how the Android bridge works under the hood, but it looks like it sometime expects to return a string (hence would not throw an exception if parsing fails) or objects.

**Comment by jpmonette** [comment]
@trooper2013 
Added some additional changes to the code and couple of fixes. I replaced our `react-native-force` dependency internally to depend on this fork and it seems to be going fine - will see if anything comes up.

Anything that has been added feature-wise to the `dev` branch has not been implemented yet, but will need to be integrated.

Please, let's connect to know what are the next steps to hopefully make this part of the next release.

**Comment by trooper2013** [comment]
> @trooper2013
> Added some additional changes to the code and couple of fixes. I replaced our `react-native-force` dependency internally to depend on this fork and it seems to be going fine - will see if anything comes up.
> 
> Anything that has been added feature-wise to the `dev` branch has not been implemented yet, but will need to be integrated.
> 
> Please, let's connect to know what are the next steps to hopefully make this part of the next release.

@jpmonette  Thanks. We will review  in the next few days.

**Comment by jpmonette** [comment]
Quick reminder here to know who'd be able to review so it doesn't go too much out of date as its been almost 2 months since last response!

cc @wmathurin @bhariharan

**Comment by AlkanV** [comment]
Hello, any update on this?

**Comment by bhariharan** [comment]
@jpmonette @AlkanV We're currently getting ready to add support for `iOS 14` and `Android 11`. Once that is complete and we release `Mobile SDK 8.3`, we will review this PR and provide feedback in a few weeks. Thank you for your patience!

**Comment by wmathurin** [comment]
We have limited bandwidth for our next release (8.3 - soon to be released). We might consider that work for our next major release (9.0 - end of year).

**Comment by jpmonette** [comment]
@wmathurin Any thoughts if this will be squeezed in 9.0 or is this postponed further?

**Comment by wmathurin** [comment]
Have you successfully run the tests on iOS and Android? Instructions are in https://github.com/forcedotcom/SalesforceMobileSDK-ReactNative/blob/dev/test/README.md

**Comment by wmathurin** [comment]
Besides running all the iOS and Android tests, you should be able to run our template apps:
* clone the [templates](https://github.com/forcedotcom/SalesforceMobileSDK-Templates) repo.
* go to the ReactNativeTemplate directory
* edit package.json to point to your fork/branch of SalesforceMobileSDK-ReactNative by changing this [line](https://github.com/forcedotcom/SalesforceMobileSDK-Templates/blob/dev/ReactNativeTemplate/package.json#L19)
* now run installios.js and installandroid.js
* you should be able to successfully generate/build/run the iOS and Android basic template app.
* you should repeat the same steps for the other template apps: ReactNativeDeferredTemplate and MobileSyncExplorerReactNative.

**Comment by wmathurin** [comment]
@jpmonette Thanks for your PR. I will close your PR.
I'm bringing most of your changes in https://github.com/forcedotcom/SalesforceMobileSDK-ReactNative/pull/212
(And I also made small changes to the [templates](https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/2143) and [android](https://github.com/forcedotcom/SalesforceMobileSDK-Templates/pull/294) repo to work with the ported library).

**Pull request closed without merging**